src/Memory/MemoryInfoProducer.cpp

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "Memory/MemoryInfoProducer.h"
+
+#include <chrono>
+#include <thread>
+
+#include "MemoryUtils.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/Profiling.h"
+#include "capture.pb.h"
+
+namespace orbit_memory {
+
+using orbit_grpc_protos::TotalMemoryUsageInfo;
+
+void MemoryInfoProducer::Start() {
+  *exit_requested_ = false;
+  thread_ = std::make_shared<std::thread>(&MemoryInfoProducer::Run, this);
+}
+
+void MemoryInfoProducer::Stop() {
+  *exit_requested_ = true;
+  if (thread_ != nullptr && thread_->joinable()) {
+    thread_->join();
+  }
+  thread_.reset();
+}
+
+void MemoryInfoProducer::Run() {
+  pthread_setname_np(pthread_self(), "MemoryInfoProducer::Run");
+
+  CHECK(listener_ != nullptr);
+  ProduceTotalMemoryUsageInfoAndSendToListener();
+}
+
+void MemoryInfoProducer::ProduceTotalMemoryUsageInfoAndSendToListener() {
+  std::optional<TotalMemoryUsageInfo> total_memory_usage_info;
+  uint64_t current_timestamp_ns = orbit_base::CaptureTimestampNs();
+  while (!(*exit_requested_)) {
+    total_memory_usage_info = GetTotalMemoryUsageInfo(current_timestamp_ns);
+    if (total_memory_usage_info.has_value()) {
+      listener_->OnTotalMemoryUsageInfo(total_memory_usage_info.value());
+    }
```

Comment:
I am not sure if this is already possible (there was a similar discussion with @ronaldfw). It would be nice, to notify the UI in case of an error. If that is not yet possible (which I assume is the case), would be nice if you could create a bug and add a TODO here.

Reply:
Thanks for the suggestion. Should I add it here or in the `GetSystemMemoryUsage()` method (the `GetTotalMemoryUsageInfo()` method is now renamed as `GetSystemMemoryUsage()`) where we generate an error if fails to read the `/proc/meminfo` file? I feel like it is a little bit strange as there is no error mentioned in this file.

Reply:
If we would have that communication to the UI, it would be added here (something like `listener_->OnError`, or `listener_->OnMetaInfo`). So I would add the TODO here.
Also note that this is totally optional. I was just wondering, as I remembered the discussion on other places.